webrender/src/frame_builder.rs

```diff
@@ -1041,36 +856,90 @@ impl FrameBuilder {
             if let Some(sc_index) = self.stacking_context_stack.last() {
                 let stacking_context = &self.stacking_context_store[sc_index.0];
                 if stacking_context.composite_ops.count() > 0 {
-                    render_mode = FontRenderMode::Alpha;
+                    normal_render_mode = FontRenderMode::Alpha;
                 }
             }
         }
 
+        // Shadows never use subpixel AA, but need to respect the alpha/mono flag
+        // for reftests.
+        let shadow_render_mode = match self.config.default_font_render_mode {
+            FontRenderMode::Subpixel | FontRenderMode::Alpha => FontRenderMode::Alpha,
+            FontRenderMode::Mono => FontRenderMode::Mono,
+        };
+
         let prim = TextRunPrimitiveCpu {
-            run: TextRun {
-                font_key,
-                logical_font_size: size,
-                glyph_range,
-                glyph_count,
-                glyph_instances: Vec::new(),
-                glyph_options,
-                render_mode: render_mode,
-                offset: LayerVector2D::zero(),
-            },
+            font_key,
+            logical_font_size: size,
+            glyph_range,
+            glyph_count,
+            glyph_instances: Vec::new(),
+            glyph_options,
+            normal_render_mode,
+            shadow_render_mode,
+            offset: LayerVector2D::zero(),
             color: *color,
         };
 
-        if self.text_shadow_builder.has_shadows() {
-            self.text_shadow_builder.push_text(prim,
-                                               rect,
-                                               local_clip.clone(),
-                                               clip_and_scroll);
-        } else if color.a > 0.0 {
+        // Text shadows that have a blur radius of 0 need to be rendered as normal
+        // text elements to get pixel perfect results for reftests. It's also a big
+        // performance win to avoid blurs and render target allocations where
+        // possible. For any text shadows that have zero blur, create a normal text
+        // primitive with the shadow's color and offset. These need to be added
+        // *before* the visual text primitive in order to get the correct paint
+        // order. Store them in a Vec first to work around borrowck issues.
+        // TODO(gw): Refactor to avoid having to store them in a Vec first.
+        let mut fast_text_shadow_prims = Vec::new();
+        for shadow_prim_index in &self.shadow_prim_stack {
+            let shadow_metadata = &self.prim_store.cpu_metadata[shadow_prim_index.0];
+            let shadow_prim = &self.prim_store.cpu_text_shadows[shadow_metadata.cpu_prim_index.0];
+            if shadow_prim.shadow.blur_radius == 0.0 {
```

Comment:
somewhat unfortunate that we already have the same loop at the start of the function, maybe we could combine them?